File: hex_terrain_editor/editor/eyedropper_tool.py
"""Eyedropper tool for sampling terrain from an existing hex."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from ..utils.hex_math import pixel_to_hex

if TYPE_CHECKING:
    from ..model.hex_grid import HexGrid


class EyedropperTool:
    """Click to sample elevation and climate from a hex."""

    def __init__(self, grid: "HexGrid", on_sample: Any = None) -> None:
        self.grid = grid
        self.on_sample = on_sample

    def activate(self) -> None:
        pass

    def deactivate(self) -> None:
        pass

    def on_press(self, world_x: float, world_y: float) -> None:
        q, r = pixel_to_hex(world_x, world_y, self.grid.hex_size)
        cell = self.grid.get(q, r)
        if cell is not None and self.on_sample:
            self.on_sample(cell.elevation, cell.climate, list(cell.decorators))

    def on_move(self, world_x: float, world_y: float) -> None:
        pass

    def on_release(self, world_x: float, world_y: float) -> None:
        pass

File: src/editor/eyedropper_tool.rs
use crate::hex_math;
use crate::model::hex_grid::HexGrid;
use crate::model::terrain_types::{Climate, Elevation};

pub struct EyedropperTool {
    /// The most recently sampled elevation and climate, if any.
    pub sampled: Option<(Elevation, Climate)>,
}

impl EyedropperTool {
    pub fn new() -> Self {
        Self { sampled: None }
    }

    pub fn on_press(&mut self, world_x: f32, world_y: f32, grid: &HexGrid) {
        let (q, r) = hex_math::pixel_to_hex(world_x, world_y, grid.hex_size);
        if let Some(cell) = grid.get(q, r) {
            self.sampled = Some((cell.elevation, cell.climate));
        }
    }
}

impl Default for EyedropperTool {
    fn default() -> Self {
        Self::new()
    }
}
